Restart runs

EtherealMC-Bit opened the pull request.

nruof committed:
Add checkpoint_diagnostic wrapper and init_restart_util

nruof committed:
Make wrapper for the checkpoint diagnostic

nruof committed:
Text cleanup

EtherealMC-Bit committed:
Merge branch 'memaster' into nruof/restart_runs

EtherealMC-Bit committed:
Added Nick's test file and function for restarting a run

EtherealMC-Bit committed:
Ensure that the checkpoint file starts with the prefix

EtherealMC-Bit committed:
Used mwxrun to initialize the simulation

EtherealMC-Bit committed:
Use diode_setup to set up the simulation

EtherealMC-Bit committed:
Added a function to clean old checkpoint directories

EtherealMC-Bit committed:
Added the option to force vs try a restart

roelof-groenewald commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -0,0 +1,52 @@
+
+        mwxrun.simulation.amr_restart = os.path.join("diags", checkpoint)
+
+    except RuntimeError:
+        logger.error("There was a error while trying to restart from the checkpoint!")
+        traceback.print_exc()
+        raise
```
In this case we'd want to try and restart from the next checkpoint file, right? So instead of raising the exception I would say we should remove the checkpoint that failed and rerun the function. Then only raise the exception if there are no more checkpoint files left over AND `force=True` (if `force=False` the simulation should just start from the beginning).

roelof-groenewald replied:
Actually, this will be more difficult since the restart won't actually happen here, so this `RuntimeError` won't catch failed restarts. So saying there was an error in trying to restart is not really correct, the error would be in trying to write `mwxrun.simulation.amr_restart`. We should think some more about how we would be able to catch failed restarts.

EtherealMC-Bit replied:
Will we just have to do that from somewhere in the input script?

roelof-groenewald replied:
I think we will probably need to wrap the `mwxrun.init_run()` functionality in a try statement.

roelof-groenewald replied:
The cleanest would probably be to edit that function (`mwxrun.init_run()`) to take an argument for whether or not the run should be restarted and have it call the `init_restart_util.run_restart()` function. Could that work?

EtherealMC-Bit replied:
That sounds like a clean solution, I will test that.

peterscherpelz replied:
@roelof-groenewald It's also worth noting we may not need the functionality of deleting restarts / trying the next one back. Ideally restarts are written to temp files and then moved with an atomic operation to the correct filename only once complete. If AMReX does this properly (I haven't checked), we won't have the issues we did with Warp where a run killed while checkpointing would have the files present but incomplete.

peterscherpelz commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -0,0 +1,52 @@
+
+def run_restart(checkpoint_prefix, force=False):
+
+    logger.info(f"Attempting to " + ("force a " if force else "") + "restart from the most recent checkpoint starting with '{checkpoint_prefix}'")
```
Break line to 80 characters

peterscherpelz replied:
And in other logger statements below.

peterscherpelz commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -0,0 +1,52 @@
+            raise RuntimeError("Diags directory does not exist!")
+        else:
+            logger.warning("Diags directory does not exist!")
+            return
```
I would `return False` so that if the user wants they can check if the restart was successful or not based on this function's return.

peterscherpelz commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -0,0 +1,52 @@
+            raise RuntimeError(f"There were no checkpoint directories starting with {checkpoint_prefix}!")
+        else:
+            logger.warning(f"There were no checkpoint directories starting with {checkpoint_prefix}!")
+            return
```
```suggestion
            return False
```

peterscherpelz reviewed (commented):
@MKieburtz Nice work so far! I'm not sure what the test exactly accomplishes. Could you clean it up some and add comments? A few ideas from me:
- It would be good to have one run that goes and writes a checkpoint so the writing functionality with our diag wrapper is tested.
- We should test all restart situations: Forced restart works; unforced restart works; test that a forced restart w/ no restart files raises an exception; test that an unforced restart properly runs a new simulation.
- For restarts, check that the step count is indeed correct after a few more steps have been run.

peterscherpelz commented:
@MKieburtz Also, please (a) merge in the new memaster and (b) change out of draft mode.

EtherealMC-Bit committed:
Merge branch 'memaster' into nruof/restart_runs

EtherealMC-Bit committed:
Broke up logger lines so they are less than 80 characters long

EtherealMC-Bit committed:
Changed the current test to just check if checkpoints were created.

EtherealMC-Bit committed:
Renamed some files

EtherealMC-Bit committed:
Fixed broken import after rename

EtherealMC-Bit committed:
Surround mwxrun.init_run() with a try-catch to catch failed restarts

roelof-groenewald commented on mewarpx/mewarpx/diags_store/checkpoint_diagnostic.py:
```diff
@@ -0,0 +1,45 @@
+        Arguments:
+            diag_steps (int): Run the diagnostic with this period.
+                Also plot on this period if enabled.
+            grid (picmi grid): The grid to be passed into the picmi checkpoint diagnostic.
```
Why is grid an optional parameter? It doesn't look like it is used anywhere.

EtherealMC-Bit replied:
Ah yep, it's not necessary with the current Picmi checkpoint.

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,27 +103,36 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
             raise RuntimeError(
                 "Attempted to initialize the mwxrun class multiple times.")
+        try:
+            if restart:
+                self.simulation.amr_restart = os.path.join("diags", checkpoint)
```
The `run_restart` function should be called here to use that functionality to find the most recent checkpoint file. So I'd say passing the `checkpoint_prefix` (which should have a standard default value for most / all of our runs) as an optional parameter to `init_run` would be the way to go. Then that `checkpoint_prefix` value is passed to `run_restart` to find which checkpoint to restart from.

EtherealMC-Bit replied:
Ok makes sense

roelof-groenewald commented on mewarpx/mewarpx/diags_store/checkpoint_diagnostic.py:
```diff
@@ -0,0 +1,45 @@
+            diag_steps (int): Run the diagnostic with this period.
+                Also plot on this period if enabled.
+            grid (picmi grid): The grid to be passed into the picmi checkpoint diagnostic.
+            name (str): The name of the diagnostic to be passed into the
+                picmi checkpoint diagnostic.
```
For `name` we should pick a default value, maybe `"chkpoint"`, like you used below. Then that can also be the default value for `checkpoint_prefix` in `init_run`.

EtherealMC-Bit committed:
Removed grid parameter from checkpoint_diagnostic

EtherealMC-Bit committed:
mwxrun.init_run() uses the init_restart_util to find a checkpoint dir

EtherealMC-Bit committed:
Added WIP test for restarting from checkpoints

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,27 +103,40 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
-
-        self.initialized = True
+        try:
+            if restart:
+                init_restart_util.run_restart(checkpoint_prefix, force_restart)
```
We should follow the strategy used in `metools` here, where by default a run would restart if there are restart files available but there aren't any it will still happily start from the beginning. However, if `restart=False`, the run should always start from the beginning and if `restart=True` it should throw an error if it cannot restart. To achieve this, the default value for `restart` can be `None` and the logic below should (hopefully work):
```suggestion
            if restart != False:
                init_restart_util.run_restart(checkpoint_prefix, restart)
```

MKieburtz replied:
Doesn't this remove the possibility of doing a fresh run if no checkpoint files are found? I'll tag @peterscherpelz because he mentioned this to me as something to include calling it 'trying' a restart. Whereas 'forcing' a restart is the behavior you suggest above for `restart=True` - throwing and error and exiting if there is a problem with restarting.

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,27 +103,40 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
 
 
-    def init_run(self):
+    def init_run(self, restart=False, force_restart=False, checkpoint_prefix="chkpoint"):
```
```suggestion
    def init_run(self, restart=None, checkpoint_prefix="chkpoint"):
```

EtherealMC-Bit committed:
Changed the parameter for mwxrun.init_run(), and checked for correct behavior

EtherealMC-Bit committed:
Added the option to run for n steps after restarting

EtherealMC-Bit committed:
Remove add particles code while the functionality for it is not merged

EtherealMC-Bit committed:
Added test for running for additional steps, and fixed another test

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,27 +103,40 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
-        self.initialized = True
+        try:
+            if restart != False:
+                init_restart_util.run_restart(checkpoint_prefix, restart is not None, additional_steps)
```
Why do `restart is not None` instead of just `restart`?

EtherealMC-Bit replied:
That was in order to not have that parameter have possible values of `True` or `None`, because that seems strange, so I converted it to a boolean.

roelof-groenewald commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -0,0 +1,91 @@
+
+    if additional_steps is None:
+        mwxrun.simulation.max_steps = max_steps - checkpoint_step
+        print("MAX STEPS ", mwxrun.simulation.max_steps)
```
We should use the logger here as well, no?

EtherealMC-Bit replied:
Yes that was a debug print to make sure things were working correctly, if you think it's useful information to print I'll change it to use logging

roelof-groenewald commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -0,0 +1,91 @@
+                f"The checkpoint directory was created at a later step ({checkpoint_step}) than the current max steps ({max_steps})!"
+            )
+    else:
+        mwxrun.simulation.max_steps = additional_steps
```
Just to be sure, how does this work if the checkpoint happened at step x? Would it then run up to step = additional_steps or step = (x + additional_steps)?

EtherealMC-Bit replied:
it runs to x + additional_steps. `mwxrun.simulation.max_steps` is the number of steps to run, not the step at which to stop.

EtherealMC-Bit committed:
Better message for printing the max step

EtherealMC-Bit committed:
Merge memaster into nruof/restart_runs

EtherealMC-Bit committed:
Add checkpoint file

EtherealMC-Bit committed:
Fix typo and print stack trace

roelof-groenewald commented on mewarpx/tests/diags/chkpoint00005/Level_0/Bx_fp_H:
```diff
@@ -0,0 +1,41 @@
+2
```
All these checkpoint files should go in `mewarpx/tests/test_files/` inside an appropriately named folder please.

EtherealMC-Bit committed:
Move checkpoint files

EtherealMC-Bit committed:
Add option to search a specific directory for checkpoints

EtherealMC-Bit committed:
Debugging circleci failure

EtherealMC-Bit committed:
fixed syntax that I didn't test

EtherealMC-Bit committed:
Maybe fixed the issue

roelof-groenewald commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -0,0 +1,92 @@
+            )
+            return False
+
+    checkpoint = checkpoints[-1:][0]
```
Is there a reason to do the above instead of
```suggestion
    checkpoint = checkpoints[-1]
```

roelof-groenewald commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -0,0 +1,92 @@
+                f"only rerun step {checkpoint_step}."
+            )
+        if mwxrun.simulation.max_steps < 0:
+            raise RuntimeError(
+                f"The checkpoint directory was created at a later step ({checkpoint_step}) than the current max steps ({max_steps})!"
+            )
```
```suggestion
            raise RuntimeError(
                "The checkpoint directory was created at a later step "
                f"({checkpoint_step}) than the current max steps ({max_steps})!"
            )
```

roelof-groenewald commented on mewarpx/mewarpx/diags_store/checkpoint_diagnostic.py:
```diff
@@ -0,0 +1,43 @@
+
+
+class CheckPointDiagnostic(WarpXDiagnostic):
+    def __init__(self, diag_steps, name, write_dir=".", **kwargs):
```
```suggestion
    def __init__(self, diag_steps, name="chkpoint", write_dir=".", **kwargs):
```

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,23 +104,41 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
 
 
-    def init_run(self):
+    def init_run(self, restart=None, checkpoint_dir="diags", checkpoint_prefix="chkpoint", additional_steps=None):
```
```suggestion
    def init_run(self, restart=None, checkpoint_dir="diags", 
                 checkpoint_prefix="chkpoint", additional_steps=None):
```

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,23 +104,41 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
-        self.initialized = True
+        try:
+            if restart != False:
+                init_restart_util.run_restart(
+                    checkpoint_dir, checkpoint_prefix,
+                    restart is not None, additional_steps
+                )
```
Seeing as the `run_restart` function returns a boolean of whether the simulation will be restarted we can use it later for the appropriate logging statement
```suggestion
                restart = init_restart_util.run_restart(
                    checkpoint_dir, checkpoint_prefix,
                    restart is not None, additional_steps
                )
```

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,23 +104,41 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
+            self.initialized = True
+
+        except RuntimeError:
+            if restart != False:
```
```suggestion
            if restart:
```

EtherealMC-Bit replied:
I modified this a little to account for the case where `init_restart_util` returned True, but there was an exception after that.

KZhu-ME replied:
If we create a force_restart variable like in my previous comment, we can just use the following here.

```suggestion
            if force_restart:
```

roelof-groenewald reviewed (commented):
This is looking good. I left a few more comments for small changes. After those are addressed I'll be ready to approve the PR.

EtherealMC-Bit committed:
Changes from code review

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,23 +104,50 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
+        except RuntimeError:
+            if restart != False:
+                if restart_success:
+                    logger.errro(
```
Typo:
```suggestion
                    logger.error(
```

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,23 +104,50 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
+                        f"a checkpoint starting with {checkpoint_prefix} "
+                        "but there was an error initializing WarpX!"
+                    )
+                else:
+                    logger.error(
+                        "There was an error restarting from "
+                        f"a checkpoint starting with {checkpoint_prefix}!"
+                    )
```
This will be triggered if `restart_success = False` (for example there are no checkpoint files available), in which case the appropriate error message would be: `"There was an error initializing the run!"`. So I would just use that same error message here (which also means that you don't really need `restart` and `restart_success`).

EtherealMC-Bit committed:
Fix typo

Co-authored-by: Roelof Groenewald <40245517+roelof-groenewald@users.noreply.github.com>

KZhu-ME commented on mewarpx/mewarpx/diags_store/checkpoint_diagnostic.py:
```diff
@@ -0,0 +1,43 @@
+
+
+class CheckPointDiagnostic(WarpXDiagnostic):
+    def __init__(self, diag_steps, name="chkpoint", write_dir=".", **kwargs):
```
```suggestion
    def __init__(self, diag_steps, name="checkpoint", write_dir=".", **kwargs):
```
Small thing but I think that if the abbreviation is just going to be 2 letters shorter than the full word, we might as well use the full word to be more descriptive.

EtherealMC-Bit replied:
ok agreed

KZhu-ME commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,23 +104,50 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
 
-    def init_run(self):
+    def init_run(self, restart=None, checkpoint_dir="diags",
+                 checkpoint_prefix="chkpoint", additional_steps=None):
```
```suggestion
                 checkpoint_prefix="checkpoint", additional_steps=None):
```

KZhu-ME commented on mewarpx/mewarpx/utils_store/init_restart_util.py:
```diff
@@ -0,0 +1,93 @@
+                print(f"Removing old checkpoint file {d}")
+                shutil.rmtree(os.path.join(directory, d))
+
+def run_restart(checkpoint_directory="diags", checkpoint_prefix="chkpoint", force=False, additional_steps=None):
```
```suggestion
def run_restart(checkpoint_directory="diags", checkpoint_prefix="checkpoint", force=False, additional_steps=None):
```

KZhu-ME commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,23 +104,50 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
-        self.lev = _libwarpx.libwarpx.warpx_finestLevel()
-
-        self.initialized = True
+        try:
+            restart_success = None
+            if restart != False:
+                restart_success = init_restart_util.run_restart(
+                    checkpoint_dir, checkpoint_prefix,
+                    restart is not None, additional_steps
+                )
```
Passing `restart is not None` to `run_restart` seems a bit weird, and doesn't immediately convey what is actually happening here. I think it'd be helpful and make the code easier to read if you created another variable called `force_restart` like the following. This will also help to simplify the `if restart != False:` logic in the next section. 

```suggestion
        # force_restart is False by default unless restart is set to True
        force_restart = restart if isinstance(restart, bool) default False
        try:
            restart_success = None
            if restart != False:
                restart_success = init_restart_util.run_restart(
                    checkpoint_dir, checkpoint_prefix,
                    force_restart, additional_steps
                )
```

EtherealMC-Bit replied:
After looking a little, it turns out `None` is falsy in python so I can just do `bool(restart)`

EtherealMC-Bit replied:
I could do `force_restart = True if bool(restart) else False` but this doesn't feel necessary to me, because that's what `bool(restart)` means anyway. Let me know what you think.

KZhu-ME replied:
`bool(restart)` works. I think the advantage of making a second variable is just making it clear that you're passing whether or not to force a restart, no whether or not you should restart.

KZhu-ME commented:
Just a few small changes, overall looks good.

EtherealMC-Bit committed:
Corrected the logic for what error to print if init_restart_util returns False

EtherealMC-Bit committed:
A few suggested changes

- Renamed the default checkpoint name to 'checkpoint'
- Made the logic of passing force to init_restart_util more clear
- Renamed the uploaded checkpoint

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -103,23 +104,41 @@ def init_timestep(self, V_anode, DT=None, CFL_factor=None, V_grid=5):
+            self.initialized = True
+
+        except RuntimeError:
+            if not restart:
```
Wouldn't `restart` be `True` here if `init_restart_util.run_restart()` found a checkpoint file to restart from? So then this line should be the following, no?
```suggestion
            if restart:
```

EtherealMC-Bit replied:
Ah yep

EtherealMC-Bit committed:
Fix if

roelof-groenewald reviewed (approved):
Looks good. Thanks!

PhilMiller commented on mewarpx/mewarpx/diags_store/checkpoint_diagnostic.py:
```diff
@@ -0,0 +1,43 @@
+
+
+class CheckPointDiagnostic(WarpXDiagnostic):
+    def __init__(self, diag_steps, name="checkpoint", write_dir=".", **kwargs):
```
Rather than have the default value string literal `"checkpoint"` in multiple places, please name a variable somewhere and use that in both sites, writing and reading checkpoints, so it's clear they're a matched pair.

PhilMiller commented:
Code generally looks good

EtherealMC-Bit committed:
Created a common variable for the default name of checkpoints

EtherealMC-Bit committed:
Merge memaster into nruof/restart_runs

KZhu-ME reviewed (approved):
Looks good!

EtherealMC-Bit committed:
Made the meaning of the bool(restart) statement clearer
